### PR TITLE
Various fixes for Samesite handling in sessions, iframes

### PR DESCRIPTION
**Pull request by diosmosis**
Fix couple issues w/ samesite handling in session, make sure session start has correct value, write header where needed, and use None for opt out frame so the session ID is sent when embedding the iframe.

fix https://github.com/matomo-org/matomo/issues/15513

### CODE REVIEW

File: core/Session.php [hunk 1]
```diff
@@ -190,7 +190,14 @@ public static function getSameSiteCookieValue()
     {
         $config = Config::getInstance();
         $general = $config->General;
-        if (!empty($general['enable_framed_pages']) && ProxyHttp::isHttps()) {
+
+        $module = Common::getRequestVar('module', false);
```
Comment (tsteur) [inline core/Session.php]: just fyi you can use `Piwik::getModule()` and `Piwik::getAction()`. not too important though

Comment (diosmosis) [reply core/Session.php]: modified

File: js/piwik.js [hunk 2]
```diff
@@ -3290,7 +3290,8 @@ if (typeof window.Piwik !== 'object') {
                     (msToExpire ? ';expires=' + expiryDate.toGMTString() : '') +
                     ';path=' + (path || '/') +
                     (domain ? ';domain=' + domain : '') +
-                    (isSecure ? ';secure' : '');
+                    (isSecure ? ';secure' : '') +
+                    (isSecure ? ';SameSite=None' : ';SameSite=Lax');
```
Comment (sgiehl) [inline js/piwik.js]: Actually that might maybe cause problems in some specific browser versions that do not accept cookies with `SameSite=None` or treat them as `SameSite=Strict` instead. There's a list of incompatible browser versions. See https://www.chromium.org/updates/same-site/incompatible-clients

**Review by sgiehl — CHANGES_REQUESTED** [review]
Looked through the changes and if I unterstand everything correct it looks mostly good so far. But I'm wondering if those lines need an adjustment as well:

https://github.com/matomo-org/matomo/blob/109926dd5a65244406bede12b897d59a23803d96/core/Cookie.php#L438-L458

First point here is that Chrome is split up into `Chrome`, `Chrome Mobile`, `Chrome Webview`, ... and many browsers are based on Chrome/Chromium and if they update their base to Chrome 80, they will require the correct SameSite values. So basically it's not good to only check for Browser with the name `Chrome` here. It would be better to check the browser family instead. So something like:

```php
$browserFamily = \DeviceDetector\Parser\Client\Browser::getBrowserFamily($deviceDetector->getClient('short_name'))
$isChrome = $browserFamily === 'Chrome';
```

Also Firefox and Edge announced to support the SameSite cookies. Won't they restrict `SameSite=None` without `Secure` as well? Not sure if the check for `Chrome` makes sense here at all.

**Comment by diosmosis** [comment]
great points @sgiehl , I'll them asap.

**Comment by tsteur** [comment]
> Also Firefox and Edge announced to support the SameSite cookies. Won't they restrict SameSite=None without Secure as well? Not sure if the check for Chrome makes sense here at all.

Just fyi they support samesite already for a while. What changes for Chrome is the default when no samesite attribute is specified, then Lax is applied, and when it says None it must be secure. AFAIK Firefox etc that's not the case just yet. We could probably apply same behaviour across browsers already though. Just means opt out might be broken for few more sites on http but might not be too many.

**Comment by diosmosis** [comment]
Updated to always add SameSite=Lax in JS. Seems to work for chrome, firefox & safari. Didn't use old versions of safari of course, though those have issues w/ None only.

**Pull request merged**